### ISSUE TITLE
Unexplained 1 Second Delay When Connecting

### ISSUE THREAD

**Issue by Carter12s** [eclipse/paho.mqtt.rust#133]
This could just be operator error, but I'm scratching my head over this one...

Relevant chunk of code:
```    
    let start_of_create = std::time::Instant::now();
    let create_opts = paho_mqtt::CreateOptionsBuilder::new()
        .mqtt_version(paho_mqtt::MQTT_VERSION_5)
        .server_uri("tcp://localhost:1883")
        .finalize();
    let mqtt_client = match paho_mqtt::AsyncClient::new(create_opts) {
        Ok(client) => client,
        Err(err) => {
            panic!(
                "Failure creating paho client, options are wrong. Err:{:?}",
                err
            );
        }
    };
    debug!("It took {:?} to create client", start_of_create.elapsed());

    // Connect the client (failure to connect we kill process)
    let conn_opts = paho_mqtt::ConnectOptionsBuilder::new()
        .keep_alive_interval(KEEP_ALIVE)
        .mqtt_version(paho_mqtt::MQTT_VERSION_5)
        .clean_start(true) // specifically for MQTT v5
        .finalize();
    let result = mqtt_client.connect(conn_opts).wait_for(MQTT_TIMEOUT);

    debug!("It took {:?} until client was connected", start_of_create.elapsed())
```

Being run with a local mosquitto broker. Produces log output:

```
[2021-08-20T00:57:37.462Z DEBUG glia_mqtt] It took 193.801µs to create client
[2021-08-20T00:57:37.462Z DEBUG paho_mqtt::async_client] Connecting handle: 0x5621b403fe20
[2021-08-20T00:57:37.462Z DEBUG paho_mqtt::async_client] Connect options: ConnectOptions { copts: MQTTAsync_connectOptions { struct_id: [77, 81, 84, 67], struct_version: 8, keepAliveInterval: 20, cleansession: 0, maxInflight: 65535, will: 0x0, username: 0x0, password: 0x0, connectTimeout: 30, retryInterval: 0, ssl: 0x0, onSuccess: None, onFailure: None, context: 0x0, serverURIcount: 0, serverURIs: 0x0, MQTTVersion: 5, automaticReconnect: 0, minRetryInterval: 1, maxRetryInterval: 60, binarypwd: MQTTAsync_connectOptions__bindgen_ty_1 { len: 0, data: 0x0 }, cleanstart: 1, connectProperties: 0x0, willProperties: 0x0, onSuccess5: None, onFailure5: None, httpHeaders: 0x0, httpProxy: 0x0, httpsProxy: 0x0 }, data: ConnectOptionsData { will: None, ssl: None, user_name: None, password: None, server_uris: StringCollection { c_coll: [], c_mut_coll: [], data: StringCollectionData { coll: [] } }, props: None, will_props: None, http_headers: None, http_proxy: None, https_proxy: None } }
[2021-08-20T00:57:38.463Z DEBUG paho_mqtt::token] Token v5 success! 0x5621b4037870, 0x7f2e8cf65d70
[2021-08-20T00:57:38.463Z DEBUG paho_mqtt::token] Token completed with code: 0
[2021-08-20T00:57:38.463Z DEBUG paho_mqtt::token] Expecting server response for: Connect
[2021-08-20T00:57:38.463Z DEBUG paho_mqtt::token] Got response: ServerResponse { rsp: Connect(ConnectResponse { server_uri: "localhost:1883", mqtt_version: 5, session_present: false }), props: Properties { cprops: MQTTProperties { count: 3, max_count: 10, length: 50, array: 0x7f2e78001d70 } }, reason_code: Success }
[2021-08-20T00:57:38.463Z DEBUG glia_mqtt] It took 1.001550197s until client was connected
```

Using wireshark to look at packets it appears as thou the broker is responding within a few micro-seconds:

![image](https://user-images.githubusercontent.com/8097919/130161699-0564baff-8f70-4bd7-841b-705897c383fa.png)

As far as I can tell changing any of the connection options has no affect on this. It looks like an internal issue with how the future is being executed, but I'm really not sure how to diagnose further.

**Comment by fpagliughi** [comment]
Hmm. Interesting. I haven't seen this behavoir when connecting. But I have seen it on disconnecting!
https://github.com/eclipse/paho.mqtt.c/issues/889

I wonder if they might be related; something in the C lib packet processing engine getting held up. 
What system are you running this on?

**Comment by Carter12s** [comment]
System details:
```
Linux 5.8.0-63-generic #71~20.04.1-Ubuntu SMP Thu Jul 15 17:46:08 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
$ cargo --version
cargo 1.52.0-nightly (90691f2bf 2021-03-16)
$ rustc --version
rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)
```
Compiling against paho_mqtt = "0.9.1"

**Comment by Carter12s** [comment]
Update:

The issue specifically only manifests when creating multiple clients, and it appears as thou client are catching responses from each-other?

A minimum reproducible example:
```
fn create_connected_client() -> paho_mqtt::AsyncClient {
    const MQTT_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(1);
    let start_of_create = std::time::Instant::now();
    let create_opts = paho_mqtt::CreateOptionsBuilder::new()
        .server_uri("tcp://localhost:1883")
        .finalize();
    let mqtt_client = paho_mqtt::AsyncClient::new(create_opts).unwrap();
    let conn_opts = paho_mqtt::ConnectOptionsBuilder::new().finalize();
    let _ = mqtt_client.connect(conn_opts).wait_for(MQTT_TIMEOUT);
    println!(
        "It took {:?} until client was connected",
        start_of_create.elapsed()
    );
    mqtt_client
}

fn main() {
    let cli_1 = create_connected_client();
    let cli_2 = create_connected_client();
    let cli_3 = create_connected_client();
    let cli_5 = create_connected_client();
    let cli_6 = create_connected_client();
}
```
Produces the following output:
```
It took 101.141428ms until client was connected
It took 1.0001531s until client was connected
It took 985.604µs until client was connected
It took 1.000180893s until client was connected
It took 986.834µs until client was connected
```
Ever even client creation takes 1 second, and odd client creations after the first complete suspiciously fast.

**Comment by Carter12s** [comment]
Attached is a a run of the previous example with tracing enabled
[trace.txt](https://github.com/eclipse/paho.mqtt.rust/files/7048592/trace.txt)

**Comment by lifeisafractal** [comment]
FYI, I have been able to reproduce this same behavior in the C++ wrapper as well. It looks like this might be due to the C lib sharing mutexes and condition variables between all clients in a process. I'm not 100% sure if this is the case, or how to fix it. More info in eclipse/paho.mqtt.cpp#365. I suggested in there that we might want to file a bug in eclise/paho.mqtt.c to consolidate discussion.
